e3sm_diags/driver/utils/dataset_xr.py

```diff
@@ -0,0 +1,921 @@
+calculated. Reference and test variables can also be derived using other
+variables from dataset files.
+"""
+from __future__ import annotations
+
+import collections
+import fnmatch
+import glob
+import os
+import re
+from typing import Callable, Dict, Literal, Tuple
+
+import cdms2
+import xarray as xr
+import xcdat as xc
+
+from e3sm_diags.derivations.derivations import (
+    DERIVED_VARIABLES,
+    DerivedVariableMap,
+    DerivedVariablesMap,
+)
+from e3sm_diags.driver import LAND_FRAC_KEY, LAND_OCEAN_MASK_PATH, OCEAN_FRAC_KEY
+from e3sm_diags.driver.utils.climo_xr import CLIMO_FREQ, CLIMO_FREQS, climo
+from e3sm_diags.logger import custom_logger
+from e3sm_diags.parameter.core_parameter import CoreParameter
+
+logger = custom_logger(__name__)
+
+
+class Dataset:
+    def __init__(
+        self,
+        parameter: CoreParameter,
+        data_type: Literal["ref", "test"],
+    ):
+        # The CoreParameter object with a list of parameters.
+        self.parameter = parameter
+
+        # The type of data for the Dataset object to store.
+        self.data_type = data_type
+
+        # The path, start year, and end year based on the dataset type.
+        if self.data_type == "ref":
+            self.root_path = self.parameter.reference_data_path
+        elif self.data_type == "test":
+            self.root_path = self.parameter.test_data_path
+        else:
+            raise ValueError(
+                f"The `type` ({self.data_type}) for this Dataset object is invalid."
+                "Valid options include 'ref' or 'test'."
+            )
+
+        # Set the `start_yr` and `end_yr` attrs based on the dataset type.
+        # Note, these attrs are different for the `area_mean_time_series`
+        # parameter.
+        if self.parameter.sets[0] in ["area_mean_time_series"]:
+            self.start_yr = self.parameter.start_yr  # type: ignore
+            self.end_yr = self.parameter.end_yr  # type: ignore
+        elif self.data_type == "ref":
+            self.start_yr = self.parameter.ref_start_yr  # type: ignore
+            self.end_yr = self.parameter.ref_end_yr  # type: ignore
+        elif self.data_type == "test":
+            self.start_yr = self.parameter.test_start_yr  # type: ignore
+            self.end_yr = self.parameter.test_end_yr  # type: ignore
+
+        # The derived variables defined in E3SM Diags. If the `CoreParameter`
+        # object contains additional user derived variables, they are added
+        # to `self.derived_vars`.
+        self.derived_vars_map = self._get_derived_vars_map()
+
+        # Whether the data is sub-monthly or not.
+        self.is_sub_monthly = False
+        if self.parameter.sets[0] in ["diurnal_cycle", "arm_diags"]:
+            self.is_sub_monthly = True
+
+    @property
+    def is_time_series(self):
+        is_ref_ts = self.parameter.ref_timeseries_input
+        is_test_ts = self.parameter.test_timeseries_input
+
+        if (self.data_type == "ref" and is_ref_ts) or (
```

Comment:
Maybe this `if` statement block can be simplified as:
```
if (self.data_type == "ref"):
    return is_ref_ts
else:
    return is_test_ts
```

Reply:
Thanks for the suggestion. I'll simplify the conditionals. I specifically added the `data_type` attr to make it easier to distinguish data types instead of using those other attrs.

Reply:
I updated the conditional to:

```python
    @property
    def is_time_series(self):
        if self.parameter.ref_timeseries_input or self.parameter.test_timeseries_input:
            return True
        else:
            return False
```

since `self.data_type` must be initialized to `"ref"` or `"test"` here: 

https://github.com/E3SM-Project/e3sm_diags/blob/c02d7015bbd1b211bdf3165d8412ace29ca23d5c/e3sm_diags/driver/utils/dataset_xr.py#L50-L59